x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java

```diff
@@ -340,4 +537,210 @@ public final void testSimpleConstantFolding() {
+                    );
+                }));
+
+                if (firstTimeSeenSignature) {
+                    List<DataType> typesWithNull = IntStream.range(0, original.types.size())
+                        .mapToObj(i -> i == finalNullPosition ? DataTypes.NULL : original.types.get(i))
+                        .toList();
+                    boolean newSignature = uniqueSignatures.add(typesWithNull);
+                    if (newSignature) {
+                        suppliers.add(new TestCaseSupplier(typesWithNull, () -> {
+                            TestCase oc = original.get();
+                            List<TypedData> data = IntStream.range(0, oc.data.size()).mapToObj(i -> {
+                                TypedData od = oc.data.get(i);
+                                if (i == finalNullPosition) {
+                                    return new TypedData(null, DataTypes.NULL, od.name);
+                                }
+                                return od;
+                            }).toList();
+                            return new TestCase(
+                                data,
+                                "LiteralsEvaluator[block=null]",
+                                entirelyNullPreservesType == false && oc.data.size() == 1 ? DataTypes.NULL : oc.expectedType,
+                                nullValue(),
+                                oc.expectedWarnings,
+                                oc.expectedTypeError
+                            );
+                        }));
+                    }
+                }
+            }
+        }
+
+        return suppliers;
+    }
+
+    /**
+     * Adds test cases containing unsupported parameter types that assert
+     * that they throw type errors.
+     */
+    protected static List<TestCaseSupplier> errorsForCasesWithoutExamples(List<TestCaseSupplier> testCaseSuppliers) {
+        for (TestCaseSupplier s : testCaseSuppliers) {
+            if (s.types == null) {
+                throw new IllegalArgumentException("types required");
+            }
+        }
+        List<TestCaseSupplier> suppliers = new ArrayList<>(testCaseSuppliers.size());
+        suppliers.addAll(testCaseSuppliers);
+
+        Set<List<DataType>> valid = testCaseSuppliers.stream().map(TestCaseSupplier::types).collect(Collectors.toSet());
+        List<Set<DataType>> validPerPosition = validPerPosition(valid);
+
+        testCaseSuppliers.stream()
+            .map(s -> s.types().size())
+            .collect(Collectors.toSet())
+            .stream()
+            .flatMap(count -> allPermutations(count))
+            .filter(types -> valid.contains(types) == false)
+            /*
+             * Skip any cases with more than one null. Our tests don't generate
+             * the full combinatorial explosions of all nulls - just a single null.
+             * Hopefully <null>, <null> cases will function the same as <null>, <valid>
+             * cases.
+             */
```

Comment:
nit: the fact that we don't generate test cases with more than 1 null would probably also be great to know in the doc string of `errorsForCasesWithoutExamples`.

Reply:
++